src/binary_to_text/binary_encoder.ml

```diff
@@ -0,0 +1,667 @@
+    add_char '\x00'
+  | Data_drop idx ->
+    write_fc buf 9;
+    write_indice buf idx
+  | Memory_copy ->
+    write_fc buf 10;
+    add_char '\x00';
+    add_char '\x00'
+  | Memory_fill ->
+    write_fc buf 11;
+    add_char '\x00'
+  | Table_init (tableidx, elemidx) ->
+    write_fc buf 12;
+    write_indice buf elemidx;
+    write_indice buf tableidx
+  | Elem_drop idx ->
+    write_fc buf 13;
+    write_indice buf idx
+  | Table_copy (idx1, idx2) ->
+    write_fc buf 14;
+    write_indice buf idx1;
+    write_indice buf idx2
+  | Table_grow idx ->
+    write_fc buf 15;
+    write_indice buf idx
+  | Table_size idx ->
+    write_fc buf 16;
+    write_indice buf idx
+  | Table_fill idx ->
+    write_fc buf 17;
+    write_indice buf idx
+  | _ -> assert false (* TODO *)
+
+and write_expr buf expr ~end_op_code =
+  List.iter (write_instr buf) expr;
+  match end_op_code with
+  | None -> Buffer.add_char buf '\x0B'
+  | Some end_op_code -> Buffer.add_char buf end_op_code
+
+let write_export buf cid ({ name; id } : Binary.export) =
+  write_string buf name;
+  Buffer.add_char buf cid;
+  write_u32_of_int buf id
+
+let write_global buf ({ typ; init; id = _ } : global) =
+  write_global_type buf typ;
+  write_expr buf init ~end_op_code:None
+
+let write_global_import buf
+  ({ Imported.modul; name; assigned_name = _; desc = mut, valtype } :
+    _ global_type Imported.t ) =
+  write_string buf modul;
+  write_string buf name;
+  Buffer.add_char buf '\x03';
+  write_valtype buf valtype;
+  write_mut buf mut
+
+let write_locals buf locals =
+  let rec compress_locals (count_types_list : (char * int) list) locals =
+    match locals with
+    | [] -> count_types_list
+    | local :: tl ->
+      let _so, vt = local in
+      let c = get_char_valtype vt in
+      let count_types_list =
+        match count_types_list with
+        | [] -> [ (c, 1) ]
+        | (ch, cnt) :: tl when ch = c -> (ch, cnt + 1) :: tl
+        | l -> (c, 1) :: l
+      in
+      compress_locals count_types_list tl
+  in
+  let count_types_list = compress_locals [] locals in
+  let count_types_list = List.rev count_types_list in
+  let len = List.length count_types_list in
+  write_u32_of_int buf len;
+  List.iter
+    (fun (char, count) ->
+      write_u32_of_int buf count;
+      Buffer.add_char buf char )
+    count_types_list
+
+let encode_section buf id encode_func data =
+  let section_buf = Buffer.create 16 in
+  encode_func section_buf data;
+  let section_len = Buffer.length section_buf in
+  if section_len = 0 then ()
+  else begin
+    Buffer.add_char buf id;
+    write_u32_of_int buf section_len;
+    Buffer.add_buffer buf section_buf
+  end
+
+(* type: section 1 *)
+let encode_types buf funcs =
+  encode_vector buf funcs (fun buf { type_f; locals = _; body = _; id = _ } ->
+      match type_f with
+      | Bt_raw (_ind, (pt, rt)) ->
+        Buffer.add_char buf '\x60';
+        write_paramtype buf pt;
+        write_resulttype buf rt
+      | _ -> assert false )
```

Comment:
You can probably get rid of this case by adding a proper type annotation. Once this is done, it should also be possible to destruct the pattern in the function parameter directly, something like: `fun buf { type_f = Bt_raw (_ind, (pt, rt)); ... } as binary ... -> ...`